core/cli/src/plugin/options.ts

```diff
@@ -59,3 +64,72 @@ export const validatePluginOptions = (logger: Logger, config: ValidPluginsConfig
 
   return invalidOptions
 }
+
+export const substituteOptionTags = (plugin: Plugin, config: ValidPluginsConfig): void => {
+  // foo.bar gets the 'bar' option set for the 'foo' plugin
+  const resolveOptionPath = (optionPath: string): unknown => {
+    const [pluginName, optionName] = optionPath.split('.', 2)
+    return (config.pluginOptions[pluginName] as OptionsForPlugin)?.options[optionName]
+  }
+
+  // throw an error if there are tags in plugin option fields to avoid circular
+  // references
+  const validateTagPath = (path: (string | number)[]) => {
+    if (path[0] === 'options' && path[1] === 'plugins') {
+      const error = new ToolKitError('cannot reference plugin options when specifying options')
+      error.details = `YAML tag referencing options used at path '${path.join('.')}'`
+      throw error
+    }
+  }
+
+  // recursively walk through the parsed config, searching for the tag
+  // identifiers we've inserted during parsing, and substitute them for
+  // resolved option values
+  const deeplySubstitute = (node: unknown, path: (string | number)[]): unknown => {
+    if (Array.isArray(node)) {
+      return node.map((item, i) => deeplySubstitute(item, [...path, i]))
+    } else if (node && typeof node === 'object') {
+      const entries = Object.entries(node)
+      // !toolkit/option will be marked by a single key within an object
+      if (entries[0]?.[0] === toolKitOptionIdent) {
```

Comment:
**suggestion**: i think having an example here in a comment of what the YAML and the resulting shape of the object in the config look like would help for people trying to understand this function